File: c++/algorithm/algorithm/list/Solution3.cpp
//
//  Solution3.cpp
//  algorithm
//
//  Created by justin.wang on 2021/5/11.
//

/**
 合并两个有序链表
 将两个升序链表合并为一个新的 升序 链表并返回。新链表是通过拼接给定的两个链表的所有节点组成的。

  

 示例 1：


 输入：l1 = [1,2,4], l2 = [1,3,4]
 输出：[1,1,2,3,4,4]
 示例 2：

 输入：l1 = [], l2 = []
 输出：[]
 示例 3：

 输入：l1 = [], l2 = [0]
 输出：[0]
  

 提示：

 两个链表的节点数目范围是 [0, 50]
 -100 <= Node.val <= 100
 l1 和 l2 均按 非递减顺序 排列

 作者：力扣 (LeetCode)
 链接：https://leetcode-cn.com/leetbook/read/tencent/x59tp7/
 来源：力扣（LeetCode）
 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
 */

#include <stdio.h>
#include <iostream>
#include <list>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* prev = new ListNode(0);
        ListNode* prevHelper = prev;
        while (l1 != nullptr && l2 != nullptr) {
            if(l1->val < l2->val) {
                prevHelper->next = l1;
                l1 = l1->next;
            } else {
                prevHelper->next = l2;
                l2 = l2->next;
            }
            prevHelper = prevHelper->next;
        }
        prevHelper->next = l1 == nullptr ? l2 : l1;
        return prev->next;
    }
};

int mainl3() {
    cout << "开始" << endl;
    ListNode list1(1);
    ListNode list2(2, &list1);
    ListNode list3(3, &list2);
    ListNode list4(4, &list3);
    ListNode list5(5, &list4);
    
    Solution so;
    
    
    return 0;
}

File: js/learn/linkedList/Solution3.js
/**
 * 2. 两数相加
给你两个 非空 的链表，表示两个非负的整数。它们每位数字都是按照 逆序 的方式存储的，并且每个节点只能存储 一位 数字。

请你将两个数相加，并以相同形式返回一个表示和的链表。

你可以假设除了数字 0 之外，这两个数都不会以 0 开头。

 

示例 1：


输入：l1 = [2,4,3], l2 = [5,6,4]
输出：[7,0,8]
解释：342 + 465 = 807.
示例 2：

输入：l1 = [0], l2 = [0]
输出：[0]
示例 3：

输入：l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]
输出：[8,9,9,9,0,0,0,1]
 

提示：

每个链表中的节点数在范围 [1, 100] 内
0 <= Node.val <= 9
题目数据保证列表表示的数字不含前导零
 */

 /**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let ans = l1;
    let moreTen = false;
    while(l1.next) {
        if(l2) {
            l1.val += l2.val;
            if(moreTen) {
                l1.val += 1;
                moreTen = false;
            }
            if(l1.val >= 10) {
                moreTen = true;
                l1.val -= 10;
            } else {
                moreTen = false;
            }
            l2 = l2.next;
        } else {
            if(moreTen) {
                l1.val += 1;
                if(l1.val >= 10) {
                    moreTen = true;
                    l1.val -= 10;
                } else {
                    moreTen = false;
                }
            } else {
            }
            
        }
        console.log('l1 va', l1, moreTen);
        l1 = l1.next;

    };
   
    if(moreTen) {
        l1.next = new ListNode(1);
    }
    
    console.log('ans = ', ans);
    return ans;
};

// let l1 = new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9)))))));
// let l2 = new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9))))

let l1 = new ListNode(1, new ListNode(9));
let l2 = new ListNode(9);

addTwoNumbers(l1, l2);